pkg/registry/service/allocator/controller/repair.go

```diff
@@ -0,0 +1,120 @@
+//
+// Can be run at infrequent intervals, and is best performed on startup of the master.
+// Is level driven and idempotent - all valid PortalIPs will be updated into the allocator
+// map at the end of a single execution loop if no race is encountered.
+//
+// TODO: allocate new IPs if necessary
+// TODO: perform repair?
+type Repair struct {
+	interval time.Duration
+	registry service.Registry
+	network  *net.IPNet
+	alloc    service.IPRegistry
+}
+
+// NewRepair creates a controller that periodically ensures that all portalIPs are uniquely allocated across the cluster
+// and generates informational warnings for a cluster that is not in sync.
+func NewRepair(interval time.Duration, registry service.Registry, network *net.IPNet, alloc service.IPRegistry) *Repair {
+	return &Repair{
+		interval: interval,
+		registry: registry,
+		network:  network,
+		alloc:    alloc,
+	}
+}
+
+// RunUntil starts the controller until the provided ch is closed.
+func (c *Repair) RunUntil(ch chan struct{}) {
+	util.Until(func() {
+		if err := c.RunOnce(); err != nil {
+			util.HandleError(err)
+		}
+	}, c.interval, ch)
+}
+
+// RunOnce verifies the state of the portal IP allocations and returns an error if an unrecoverable problem occurs.
+func (c *Repair) RunOnce() error {
+	latest, err := c.alloc.Get()
+	if err != nil {
+		return fmt.Errorf("unable to refresh the service IP block: %v", err)
+	}
+
+	ctx := api.WithNamespace(api.NewDefaultContext(), api.NamespaceAll)
+	list, err := c.registry.ListServices(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to refresh the service IP block: %v", err)
+	}
+
+	r := allocator.NewCIDRRange(c.network)
+	for _, svc := range list.Items {
+		if !api.IsServiceIPSet(&svc) {
+			continue
+		}
+		ip := net.ParseIP(svc.Spec.PortalIP)
+		if ip == nil {
+			// portal IP is broken, reallocate
+			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s is not a valid IP; please recreate", svc.Spec.PortalIP, svc.Name, svc.Namespace))
+			continue
+		}
+		switch err := r.Allocate(ip); err {
+		case nil:
+		case allocator.ErrAllocated:
+			// TODO: send event
+			// portal IP is broken, reallocate
+			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s was assigned to multiple services; please recreate", ip, svc.Name, svc.Namespace))
+		case allocator.ErrNotInRange:
+			// TODO: send event
+			// portal IP is broken, reallocate
+			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s is not within the service CIDR %s; please recreate", ip, svc.Name, svc.Namespace, c.network))
+		case allocator.ErrFull:
+			// TODO: send event
+			return fmt.Errorf("the service CIDR %s is full; you must widen the CIDR in order to create new services")
+		default:
+			return fmt.Errorf("unable to allocate portal IP %s for service %s/%s due to an unknown error, exiting: %v", ip, svc.Name, svc.Namespace, err)
+		}
+	}
+
+	service.SnapshotRange(latest, r)
+
+	if err := c.alloc.CreateOrUpdate(latest); err != nil {
```

Comment:
Yes - atomic update CAS does not write if the updated value is == to the value on the server.

----- Original Message -----
> > +			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s was
> > assigned to multiple services; please recreate", ip, svc.Name,
> > svc.Namespace))
> > +		case allocator.ErrNotInRange:
> > +			// TODO: send event
> > +			// portal IP is broken, reallocate
> > +			util.HandleError(fmt.Errorf("the portal IP %s for service %s/%s is not
> > within the service CIDR %s; please recreate", ip, svc.Name, svc.Namespace,
> > c.network))
> > +		case allocator.ErrFull:
> > +			// TODO: send event
> > +			return fmt.Errorf("the service CIDR %s is full; you must widen the CIDR
> > in order to create new services")
> > +		default:
> > +			return fmt.Errorf("unable to allocate portal IP %s for service %s/%s
> > due to an unknown error, exiting: %v", ip, svc.Name, svc.Namespace, err)
> > +		}
> > +	}
> > +
> > +	service.SnapshotRange(latest, r)
> > +
> > +	if err := c.alloc.CreateOrUpdate(latest); err != nil {
> 
> Will this properly handle no-op changes?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7704/files#r29889054